mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionResolver.java

```diff
@@ -83,4 +98,53 @@ public Mono<PostgresqlConnection> resolver(Domain domain) {
         }
     }
 
+    class PostgresConnectionPoolResolver implements PostgresConnectionResolver, Closeable {
+        private static final Logger LOGGER = LoggerFactory.getLogger(PostgresConnectionPoolResolver.class);
+
+        private final ConnectionPool connectionPool;
+
+        private final Map<String, Connection> mapConnectionLookup;
+
+        public PostgresConnectionPoolResolver(ConnectionPoolConfiguration poolConfiguration) {
+            this.connectionPool = new ConnectionPool(poolConfiguration);
+
+            this.mapConnectionLookup = new HashMap<>();
+        }
+
+        @Override
+        public Mono<? extends Connection> resolver(Domain domain) {
+            return Mono.fromCallable(() -> mapConnectionLookup.get(domain.asString()))
+                .switchIfEmpty(createPooledConnection(domain)
+                    .doOnSuccess(connection -> mapConnectionLookup.put(domain.asString(), connection)));
```

Comment:
`doOnSuccess` as I know it is async. Any chance that resolved Connection -> execute bussiness query -> release the connection (remove no Map Entry yet) -> then `doOnSuccess` adds the Map Entry (async and could be a bit late).

Then in that case the Map `mapConnectionLookup` would have a entry that points a Domain to a closed connection.